### PR TITLE
CCIP-1290 Processing Commit Roots in steps 

### PR DESCRIPTION
**Pull request by mateusz-sekara**
## Solution

### CODE REVIEW

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go [hunk 1]
```diff
@@ -211,143 +213,153 @@ func (r *ExecutionReportingPlugin) Observation(ctx context.Context, timestamp ty
...
-		if err1 != nil {
-			return nil, err1
-		}
-		return getTokensPrices(ctx, sourceFeeTokens, r.config.sourcePriceRegistry, []common.Address{r.config.sourceWrappedNativeToken})
-	})
-	getDestTokensPrices := cache.LazyFetch(func() (map[common.Address]*big.Int, error) {
-		dstTokens, err1 := r.cachedDestTokens.Get(ctx)
-		if err1 != nil {
-			return nil, err1
+	allUnexpiredNotSnoozedReports := func() []ccipdata.CommitStoreReport {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/execution_reporting_plugin.go]: I find this part of the code `allUnexpiredNotSnoozedReports = func(..)()` confusing.

Why do you need the function? Why not just do:
```
	index := 0
	for _, rep := range allUnexpiredReports {
		if r.snoozedRoots.IsSnoozed(rep.MerkleRoot) {
			lggr.Debug("Skipping snoozed root", "minSeqNr", rep.Interval.Min, "maxSeqNr", rep.Interval.Max)
		} else {
			allUnexpiredReports[index] = rep
			index++
		}
	}
	allUnexpiredNotSnoozedReports := allUnexpiredReports[:index]
```

Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go]: it also seems confusing that we mess with the elements of `allUnexpiredReports` slice
```
allUnexpiredReports[index] = rep
```

I suggest placing all this logic in a method to prevent this confusion.

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go [hunk 2]
```diff
@@ -1162,3 +1174,17 @@ func getUnexpiredCommitReports(
 	}
 	return reports, nil
 }
+
+func selectReportsToFillBatch(unexpiredReports []ccipdata.CommitStoreReport, messagesLimit int) ([]ccipdata.CommitStoreReport, int) {
+	currentNumberOfMessages := 0
+	var index int
+
+	for index = 0; index < len(unexpiredReports); index++ {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/execution_reporting_plugin.go]: you can do `for index := range unexpiredReports`

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go [hunk 3]
```diff
@@ -1162,3 +1174,17 @@ func getUnexpiredCommitReports(
 	}
 	return reports, nil
 }
+
+func selectReportsToFillBatch(unexpiredReports []ccipdata.CommitStoreReport, messagesLimit int) ([]ccipdata.CommitStoreReport, int) {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/execution_reporting_plugin.go]: can you add a comment for this function explaining how it selects the reports

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go [hunk 4]
```diff
@@ -1162,3 +1174,17 @@ func getUnexpiredCommitReports(
...
+	currentNumberOfMessages := 0
+	var index int
+
+	for index = 0; index < len(unexpiredReports); index++ {
+		currentNumberOfMessages += int(unexpiredReports[index].Interval.Max - unexpiredReports[index].Interval.Min + 1)
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/execution_reporting_plugin.go]: should we worry about casting `uint64` to `int` ?

Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go]:
```
Test_selectReportsToFillBatch
{Interval: ccipdata.CommitStoreInterval{Min: 26, Max: math.MaxUint64}},
```

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go (hunk 1 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go]: I wrapped it with function to limit the scope of all these variables. I gonna extract to function then

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go (hunk 3 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go]: Hmm, I thought it was self explanatory, also because of the specs covering all the different scenarios. Anyway, I can write sth there

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go (hunk 4 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go]: Good catch

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go [hunk 5]
```diff
@@ -1160,5 +1159,37 @@ func getUnexpiredCommitReports(
...
+	lggr.Infow("Unexpired roots", "all", len(reports), "notSnoozed", len(notSnoozedReports))
+	return notSnoozedReports, nil
+}
+
+// selectReportsToFillBatch returns the reports to fill the message limit. Single Commit Root contains exactly (Interval.Max - Interval.Min + 1) messages.
+// We keep adding reports until we reach the message limit. Please see the tests for more examples and edge cases.
+// unexpiredReports have to be sorted by Interval.Min. Otherwise, the batching logic will not be efficient,
+// because it picks messages and execution states based on the report[0].Interval.Min - report[len-1].Interval.Max range.
+// Having unexpiredReports not sorted properly will lead to fetching more messages and execution states to the memory than the messagesLimit provided.
+// However, logs from LogPoller are returned ordered by (block_number, log_index), so it should preserve the order of Interval.Min.
```
Comment (mateusz-sekara) [inline core/services/ocr2/plugins/ccip/execution_reporting_plugin.go:1181]: @connorwstein can we rely on that, or should apply additional sorting here?

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go [hunk 6]
```diff
@@ -1160,5 +1159,37 @@ func getUnexpiredCommitReports(
 	for _, acceptedReport := range acceptedReports {
 		reports = append(reports, acceptedReport.Data)
 	}
-	return reports, nil
+
+	notSnoozedReports := make([]ccipdata.CommitStoreReport, 0)
+	for _, report := range reports {
+		if r.snoozedRoots.IsSnoozed(report.MerkleRoot) {
+			lggr.Debug("Skipping snoozed root", "minSeqNr", report.Interval.Min, "maxSeqNr", report.Interval.Max)
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/execution_reporting_plugin.go:1166]: nit: i'd prefer to add a `continue` and not have the `else` part

**Review by dimkouv — APPROVED** [review]

File: core/services/ocr2/plugins/ccip/execution_reporting_plugin.go (hunk 5 above)
Comment (connorwstein) [reply core/services/ocr2/plugins/ccip/execution_reporting_plugin.go:1181]: yeah we can https://github.com/smartcontractkit/ccip/blob/0f9bf06d9a55683f4bca71d0a01672dfc6122a96/core/chains/evm/logpoller/orm.go#L378. Worth adding a comment to the commit store reader intf about that requirement actually.

**Review by connorwstein — APPROVED** [review]
